nova/virt/vmwareapi/driver.py

```diff
@@ -824,8 +824,12 @@ def check_can_live_migrate_source(self, context, instance,
             # We have to do it in the check, because the next step
             # is the pre_live_migration and that is on the destination
             # host
-            self._vmops.update_cluster_placement(context, instance,
-                                                 remove=True)
+            try:
+                self._vmops.update_cluster_placement(context, instance,
```

Comment:
This only raises, if we cleared the cache when the instance went away, right? I cannot find a reference to `vm_ref_cache_delete()` that does this.

Reply:
The cache can be cleared also by a restart or by the property-collector.
The first one is the main problem. If we roll out this patch, the nova-compute will restart and this will throw.

Reply:
Right. Would it maybe make sense to do an explicit query against the vCenter for live-migration, if the instance still exists? We probably don't want to restart nova-compute every time we have to restart a migration.

Reply:
The downsides of querying it explicitly is an additional overhead for the standard-case (VM is there, since we want to migrate the VM) for handling the exceptional case (VM is already gone, we are retrying), and the introduction of a race-condition with an ongoing live-migration. 

If I restart the whole thing while the migration is ongoing, I may get the response that the VM is there. 
If then the migration also completes before I call the `update_cluster_placement`, I'll get an `InstanceNotFound`-exception.

That is probably also the reason I didn't see the error before: I only tested restarting the compute node (and migration) early in the migration.

Reply:
I'm not arguing against handling the `InstanceNotFound`, I just wanted to make sure it's properly raised if there's no VM anymore right at the start. We would otherwise only recognize it after we run `VMOps.get_info()` through the regular status-update task.

Reply:
You're right, that this incurs and additional penalty for the standard case, but I would be willing to take this penalty to detect problem early.

Reply:
An early bail-out in case of a succeeded (or ongoing) migration is already in place here: https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/virt/vmwareapi/driver.py#L813-L814

Now I wonder if this change is even necessary, because I can't see a way how to get an `InstanceNotFound` without having a VM in the target vcenter.

Reply:
hm ... might not be necessary, you're right. Then we go back to just fixing not calling the admin-group removal.